### PR TITLE
*: support time constant push down in mocktikv

### PR DESCRIPTION
**Pull request by hanfei1991**

### CODE REVIEW

File: store/tikv/coprocessor.go [hunk 1]
```diff
@@ -42,18 +42,18 @@ func (c *CopClient) IsRequestTypeSupported(reqType, subType int64) bool {
...
+		return c.supportExpr(tipb.ExprType(subType))
 	}
 	return false
 }
 
-func supportExpr(exprType tipb.ExprType) bool {
+func (c *CopClient) supportExpr(exprType tipb.ExprType) bool {
 	switch exprType {
 	case tipb.ExprType_Null, tipb.ExprType_Int64, tipb.ExprType_Uint64, tipb.ExprType_String, tipb.ExprType_Bytes,
-		tipb.ExprType_MysqlDuration, tipb.ExprType_MysqlTime, tipb.ExprType_MysqlDecimal,
```
Comment (shenli) [inline store/tikv/coprocessor.go]: So ExprType_MysqlTime is pushed down in the current master?

File: store/tikv/coprocessor.go [hunk 2]
```diff
@@ -42,18 +42,18 @@ func (c *CopClient) IsRequestTypeSupported(reqType, subType int64) bool {
 		case kv.ReqSubTypeGroupBy, kv.ReqSubTypeBasic, kv.ReqSubTypeTopN:
 			return true
 		default:
-			return supportExpr(tipb.ExprType(subType))
+			return c.supportExpr(tipb.ExprType(subType))
 		}
 	case kv.ReqTypeDAG:
-		return true
```
Comment (shenli) [inline store/tikv/coprocessor.go]: Is this a bug fix here?

File: expression/expr_to_pb.go [hunk 3]
```diff
@@ -73,7 +74,7 @@ func (pc pbConverter) exprToPB(expr Expression) *tipb.Expr {
 	return nil
 }
 
-func (pc pbConverter) datumToPBExpr(d types.Datum) *tipb.Expr {
+func (pc pbConverter) datumToPBExpr(d types.Datum, ft *types.FieldType) *tipb.Expr {
```
Comment (shenli) [inline expression/expr_to_pb.go]: It is better to rename it to constantToPBExpr?

File: store/tikv/coprocessor.go (hunk 2 above)
Comment (hanfei1991) [inline store/tikv/coprocessor.go]: no

File: store/tikv/coprocessor.go (hunk 1 above)
Comment (hanfei1991) [inline store/tikv/coprocessor.go]: no, This check is not used.

File: expression/expr_to_pb.go (hunk 3 above)
Comment (hanfei1991) [inline expression/expr_to_pb.go]: The input is a datum but not expression.Constant

Comment (shenli) [inline expression/expr_to_pb.go]: I search in this PR for the usage and found that this function only handle Constant. So we could rename it and reduce the two parameter to one *Constant.

Comment (hanfei1991) [inline expression/expr_to_pb.go]: ok

**Comment by shenli** [comment]
LGTM

**Comment by hanfei1991** [comment]
tidb-test failed, please don't merge

File: expression/expr_to_pb.go [hunk 4]
```diff
@@ -103,6 +109,18 @@ func (pc pbConverter) datumToPBExpr(d types.Datum) *tipb.Expr {
 	case types.KindMysqlDecimal:
 		tp = tipb.ExprType_MysqlDecimal
 		val = codec.EncodeDecimal(nil, d)
+	case types.KindMysqlTime:
+		tp = tipb.ExprType_MysqlTime
+		if pc.client.IsRequestTypeSupported(kv.ReqTypeDAG, int64(tipb.ExprType_MysqlTime)) {
+			loc := pc.sc.TimeZone
+			val, err := tablecodec.EncodeValue(d, loc)
+			if err != nil {
+				log.Error(err.Error())
```
Comment (zimulala) [inline expression/expr_to_pb.go]: Please complete this error log. And `err` is enough.

**Comment by hanfei1991** [comment]
Don't merge util https://github.com/pingcap/tikv/pull/2177/files be merged

**Comment by shenli** [comment]
@hanfei1991 Can we merge this PR now?

**Comment by hanfei1991** [comment]
@shenli yes

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**